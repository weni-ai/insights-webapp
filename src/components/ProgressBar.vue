<template>
  <section class="progress-bar-container">
    <UnnnicProgressBar
      v-model="progress"
      data-test-id="progress-bar"
      inline
      :title="title"
    />
  </section>
</template>

<script>
export default {
  name: 'ProgressBar',
  props: {
    title: {
      type: String,
      default: '',
    },
    timeInterval: {
      type: Number,
      default: 50,
    },
  },
  emits: ['progress-complete'],
  data() {
    return {
      progress: 0,
      interval: null,
    };
  },
  mounted() {
    setTimeout(this.startProgressBar, 2000);
  },
  unmounted() {
    clearInterval(this.interval);
  },
  methods: {
    startProgressBar() {
      this.progress = 0;

      this.interval = setInterval(this.updateProgress, this.timeInterval);
    },
    updateProgress() {
      this.progress === 100
        ? this.$emit('progress-complete')
        : (this.progress += 1);
    },
  },
};
</script>

<style lang="scss" scoped>
.progress-bar-container {
  position: absolute;
  width: 100vw;
  height: 100vh;
  display: flex;
  justify-content: center;
  align-items: center;
  z-index: 999999;
  background-color: rgba(0, 0, 0, 0.4);
  top: 0;
  left: 0;
}
</style>
