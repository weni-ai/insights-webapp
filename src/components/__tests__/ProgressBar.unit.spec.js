import { beforeEach, describe, expect, it, vi } from 'vitest';
import { mount } from '@vue/test-utils';
import ProgressBar from '@/components/ProgressBar.vue';

describe('ProgressBar', () => {
  let wrapper;

  const createWrapper = (props = {}) => {
    return mount(ProgressBar, {
      props,
      global: {
        stubs: {
          UnnnicProgressBar: true,
        },
      },
    });
  };

  beforeEach(() => {
    vi.useFakeTimers();
    wrapper = createWrapper({ title: 'Loading...' });
  });

  afterEach(() => {
    vi.clearAllTimers();
  });

  it('Should render the progress bar with the provided title', () => {
    const progressBar = wrapper.findComponent('[data-test-id="progress-bar"]');
    expect(progressBar.exists()).toBe(true);
    expect(progressBar.props('title')).toBe('Loading...');
  });

  it('Should initialize progress at 0', () => {
    expect(wrapper.vm.progress).toBe(0);
  });

  it('Should start the progress bar after a delay of 2000ms', () => {
    vi.advanceTimersByTime(2000);
    expect(wrapper.vm.progress).toBe(0);
    expect(wrapper.vm.interval).not.toBeNull();
  });

  it('Should increment progress periodically based on timeInterval prop', () => {
    wrapper = createWrapper({ timeInterval: 100 });

    vi.advanceTimersByTime(2000);
    vi.advanceTimersByTime(100);

    expect(wrapper.vm.progress).toBe(1);

    vi.advanceTimersByTime(300);
    expect(wrapper.vm.progress).toBe(4);
  });

  it('Should emit "progress-complete" when progress reaches 100', async () => {
    vi.advanceTimersByTime(2000);
    wrapper.vm.progress = 99;

    vi.advanceTimersByTime(100);
    await wrapper.vm.$nextTick();

    expect(wrapper.vm.progress).toBe(100);
    expect(wrapper.emitted('progress-complete')).toBeTruthy();
  });

  it('Should stop interval on component unmount', () => {
    const clearIntervalSpy = vi.spyOn(window, 'clearInterval');

    wrapper.unmount();
    expect(clearIntervalSpy).toHaveBeenCalledWith(wrapper.vm.interval);
  });

  it('Matches the snapshot', () => {
    expect(wrapper.element).toMatchSnapshot();
  });
});
