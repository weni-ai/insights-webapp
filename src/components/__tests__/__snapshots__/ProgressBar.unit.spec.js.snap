// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ProgressBar > Matches the snapshot 1`] = `
<section
  class="progress-bar-container"
  data-v-84df1bb9=""
>
  <div
    class="unnnic-progress-bar primary inline"
    data-test-id="progress-bar"
    data-v-84df1bb9=""
    data-v-ac4fc1a8=""
  >
    <div
      class="main"
      data-v-ac4fc1a8=""
    >
      <div
        class="title"
        data-v-ac4fc1a8=""
      >
        Loading...
      </div>
      <div
        class="progress-bar-container"
        data-v-ac4fc1a8=""
      >
        <div
          class="progress-container"
          data-v-ac4fc1a8=""
        >
          <div
            class="bar"
            data-v-ac4fc1a8=""
            style="width: 0%;"
          >
            <!---->
          </div>
        </div>
        <div
          class="percentage"
          data-v-ac4fc1a8=""
        >
          0% 
        </div>
      </div>
      <!---->
    </div>
    <!---->
  </div>
</section>
`;
